internal/api/migrate/grafana.go

```diff
@@ -0,0 +1,245 @@
+		return err
+	}
+	*p = panel
+	return nil
+}
+
+type CurrentValue struct {
+	Value *variable.DefaultValue `json:"value,omitempty"`
+}
+
+type TemplateVar struct {
+	Name        string          `json:"name"`
+	Type        string          `json:"type"`
+	Description string          `json:"description"`
+	Label       string          `json:"label"`
+	Hide        int             `json:"hide"`
+	Sort        *int            `json:"sort,omitempty"`
+	IncludeAll  bool            `json:"includeAll"`
+	AllValue    string          `json:"allValue"`
+	Multi       bool            `json:"multi"`
+	Current     *CurrentValue   `json:"current,omitempty"`
+	Query       json.RawMessage `json:"query"`
+	json.RawMessage
+}
+
+func (v *TemplateVar) UnmarshalJSON(data []byte) error {
+	var tmp map[string]json.RawMessage
+	if err := json.Unmarshal(data, &tmp); err != nil {
+		return err
+	}
+	grafanaVariable := TemplateVar{}
+	if name, ok := tmp["name"]; ok {
+		_ = json.Unmarshal(name, &grafanaVariable.Name)
+	}
+	if t, ok := tmp["type"]; ok {
+		_ = json.Unmarshal(t, &grafanaVariable.Type)
+	}
+	if description, ok := tmp["description"]; ok {
+		_ = json.Unmarshal(description, &grafanaVariable.Description)
+	}
+	if label, ok := tmp["label"]; ok {
+		_ = json.Unmarshal(label, &grafanaVariable.Label)
+	}
+	if hide, ok := tmp["hide"]; ok {
+		_ = json.Unmarshal(hide, &grafanaVariable.Hide)
+	}
+	if sort, ok := tmp["sort"]; ok {
+		_ = json.Unmarshal(sort, &grafanaVariable.Sort)
+	}
+	if includeAll, ok := tmp["includeAll"]; ok {
+		_ = json.Unmarshal(includeAll, &grafanaVariable.IncludeAll)
+	}
+	if allValue, ok := tmp["allValue"]; ok {
+		_ = json.Unmarshal(allValue, &grafanaVariable.AllValue)
+	}
+	if multi, ok := tmp["multi"]; ok {
+		_ = json.Unmarshal(multi, &grafanaVariable.Multi)
+	}
+	if current, ok := tmp["current"]; ok {
+		if err := json.Unmarshal(current, &grafanaVariable.Current); err != nil {
+			return err
+		}
+		delete(tmp, "current")
+	}
+	if query, ok := tmp["query"]; ok {
+		if err := json.Unmarshal(query, &grafanaVariable.Query); err != nil {
+			return err
+		}
+	}
+	var err error
+	grafanaVariable.RawMessage, err = json.Marshal(tmp)
+	if err != nil {
+		return err
+	}
+	*v = grafanaVariable
+	return nil
+}
+
+func (v *TemplateVar) getDefaultValue() *variable.DefaultValue {
+	if v.Current == nil {
+		return nil
+	}
+	return v.Current.Value
+}
+
+type SimplifiedDashboard struct {
+	UID        string  `json:"uid,omitempty"`
+	Title      string  `json:"title"`
+	Panels     []Panel `json:"panels"`
+	Templating struct {
+		List []TemplateVar `json:"list"`
+	} `json:"templating"`
+}
+
+func (d *SimplifiedDashboard) UnmarshalJSON(data []byte) error {
+	tmp := &SimplifiedDashboard{}
+	type plain SimplifiedDashboard
+	if err := json.Unmarshal(data, (*plain)(tmp)); err != nil {
+		return err
+	}
+	tmp.rearrangeGrafanaPanelsWithinExpandedRows()
+	*d = *tmp
+	return nil
+}
+
+func (d *SimplifiedDashboard) rearrangeGrafanaPanelsWithinExpandedRows() {
```

Comment:
let's keep the intro comment too?

```suggestion
/**
 * This function addresses an issue we have with Grafana datamodel when it comes to migrating dashboards to Perses: When
 * a row is expanded in Grafana, its children panels are moved up in the main panels list, thus become siblings of the row.
 * When it comes to Perses migration we need to recompose the parent->children relationships. However in its current state
 * the CUE language doesn't permit us to achieve this recomposition, hence this processing in the backend code.
 *
 * So what this function does is basically the following: whenever such pattern is encountered in the panels list:
 * ...
 * row1,
 * panelA,
 * panelB,
 * panelC,
 * row2,
 * ...
 * the objects get rearranged like:
 * ...
 * row1: {
 *   panelA,
 *   panelB,
 *   panelC,
 * },
 * row2,
 * ...
 */
func (d *SimplifiedDashboard) rearrangeGrafanaPanelsWithinExpandedRows() {
```